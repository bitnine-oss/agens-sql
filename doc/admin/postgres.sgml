<!DOCTYPE book PUBLIC "-//OASIS//DTD DocBook V4.2//EN" [

<!ENTITY % version SYSTEM "version.sgml">
%version;
<!ENTITY % filelist SYSTEM "filelist.sgml">
%filelist;
<!ENTITY reference  SYSTEM "reference.sgml">

]>
<book id="postgres" lang=ko>
<title>Agens SQL 1.0 문서</title>
<bookinfo>
  <productname>Agens SQL</productname>
  <productnumber>1.0</productnumber>
</bookinfo>

<part id="admin">
 <title>Administrator</title>
  <partintro>
   <para> 
   관리자 안내서는 <productname>Agens SQL</productname> 데이터베이스 관리자에게 필요한 주제를 다룬다.
   소프트웨어 설치 및 셋업, 서버 환경 설정, 사용자와 데이터베이스 관리, 정기적인 관리에 대한 내용이 들어 있다. 
   <productname>Agens SQL</productname> 서버를 개인 용도로 쓰거나 특히 운영상 사용해야 하는 경우, 이 주제들에 익숙해야 한다.
   관리자 안내서는 처음 <productname>Agens SQL</productname> 사용시 읽으면 좋은 순서로 정리해 놓았다. 
   각 장의 내용은 독립적이기 때문에 개인의 선호에 따라 읽으면 된다. 관리자 안내서 내용은 주제 단위로 기술 되어 있으며, 
   <productname>Agens SQL</productname> 데이터베이스 시스템 기본 사용법에 익숙한 독자들이 읽을 수 있다.
</para>
  </partintro>
  &runtime;
  &config;
  &client-auth;
  &user-manag;
  &manage-ag;
  &charset;
  &maintenance;
  &backup;
  &high-availability;
  &recovery-config;
  &monitoring;
  &diskusage;
  &wal;
</part>
 <![%include-index;[&bookindex;]]>
 <![%include-xslt-index;[<index id="bookindex"></index>]]>
</book>
