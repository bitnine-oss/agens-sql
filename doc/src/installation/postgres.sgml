<!DOCTYPE book PUBLIC "-//OASIS//DTD DocBook V4.2//EN" [

<!ENTITY % version SYSTEM "version.sgml">
%version;
<!ENTITY % filelist SYSTEM "filelist.sgml">
%filelist;
<!ENTITY reference  SYSTEM "reference.sgml">

]>
<book id="postgres" lang=ko>
<title>Agens SQL 1.0 문서</title>
<bookinfo>
<productname>Agens SQL</productname>
<productnumber>1.0</productnumber>
</bookinfo>
 <part id="installation"> 
 <title>Installation</title>
 <partIntro>
   <para>Installation를 소개하는 부분입니다.</para>
 </partIntro>
 &techSheet;
 &install;
 &basicStructure;
 &databaseCreateUse;
 </part>
</book>
