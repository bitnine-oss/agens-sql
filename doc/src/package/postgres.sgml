<!DOCTYPE book PUBLIC "-//OASIS//DTD DocBook V4.2//EN" [

<!ENTITY % version SYSTEM "version.sgml">
%version;
<!ENTITY % filelist SYSTEM "filelist.sgml">
%filelist;
<!ENTITY reference  SYSTEM "reference.sgml">

]>
<book id="postgres" lang=ko>
<title>Agens SQL 1.0 문서</title>
<bookinfo>
<productname>Agens SQL</productname>
<productnumber>1.0</productnumber>
</bookinfo>
<part id="package">
 <title>Packages</title>
 <partintro>
  <para> Package를 소개하는 부분입니다. </para>
 </partintro>
 &plproxy;
 &pgbouncer;
 &skytools;
 &pgpoolII;
 &slonyI;
 &postGis;
 &pgHintPlan;
 &powa;
</part>
</book>
