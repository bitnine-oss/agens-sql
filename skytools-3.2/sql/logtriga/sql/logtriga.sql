-- init
\set ECHO none
\i logtriga.sql
\set ECHO all

create table rtest (
	id integer primary key,
	dat text
);

create table clog (
  id serial,
  op text,
  data text
);

create trigger rtest_triga after insert or update or delete on rtest
for each row execute procedure logtriga('kv',
'insert into clog (op, data) values ($1, $2)');

-- simple test
insert into rtest values (1, 'value1');
update rtest set dat = 'value2';
delete from rtest;
select * from clog; delete from clog;

-- test new fields
alter table rtest add column dat2 text;
insert into rtest values (1, 'value1');
update rtest set dat = 'value2';
delete from rtest;
select * from clog; delete from clog;

-- test field rename
alter table rtest alter column dat type integer using 0;
insert into rtest values (1, '666', 'newdat');
update rtest set dat = 5;
delete from rtest;
select * from clog; delete from clog;


-- test field ignore
drop trigger rtest_triga on rtest;
create trigger rtest_triga after insert or update or delete on rtest
for each row execute procedure logtriga('kiv',
'insert into clog (op, data) values ($1, $2)');

insert into rtest values (1, '666', 'newdat');
update rtest set dat = 5, dat2 = 'newdat2';
update rtest set dat = 6;
delete from rtest;
select * from clog; delete from clog;


-- test wrong key
drop trigger rtest_triga on rtest;
create trigger rtest_triga after insert or update or delete on rtest
for each row execute procedure logtriga('vik',
'insert into clog (op, data) values ($1, $2)');

insert into rtest values (1, 0, 'non-null');
insert into rtest values (2, 0, NULL);
update rtest set dat2 = 'non-null2' where id=1;
update rtest set dat2 = NULL where id=1;
update rtest set dat2 = 'new-nonnull' where id=2;

delete from rtest where id=1;
delete from rtest where id=2;

select * from clog; delete from clog;




