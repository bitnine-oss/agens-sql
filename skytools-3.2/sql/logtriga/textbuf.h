struct TBuf;

typedef struct TBuf TBuf;

TBuf *tbuf_alloc(int start_size);
void tbuf_free(TBuf *tbuf);
int tbuf_get_size(TBuf *tbuf);
void tbuf_reset(TBuf *tbuf);

const text *tbuf_look_text(TBuf *tbuf);
const char *tbuf_look_cstring(TBuf *tbuf);

void tbuf_append_cstring(TBuf *tbuf, const char *str);
void tbuf_append_text(TBuf *tbuf, const text *str);
void tbuf_append_char(TBuf *tbuf, char chr);

text *tbuf_steal_text(TBuf *tbuf);

void tbuf_encode_cstring(TBuf *tbuf,
		const char *str,
		const char *encoding);

void tbuf_encode_data(TBuf *tbuf,
		const uint8 *data, int len,
		const char *encoding);

