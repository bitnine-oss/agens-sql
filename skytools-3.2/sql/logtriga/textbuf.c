
#include <postgres.h>

#include "utils/elog.h"
#include "funcapi.h"
#include "mb/pg_wchar.h"
#include "parser/keywords.h"

#if 1
#define talloc(len)        malloc(len)
#define trealloc(p, len)   realloc(p, len)
#define tfree(p)           free(p)
#else
#define talloc(len)        palloc(len)
#define trealloc(p, len)   repalloc(p, len)
#define tfree(p)           pfree(p)
#endif

#include "textbuf.h"

#ifndef SET_VARSIZE
#define SET_VARSIZE(x, len) VARATT_SIZEP(x) = (len)
#endif

struct TBuf {
	text *data;
	int size;
};

static void request_avail(TBuf *tbuf, int len)
{
	int newlen = tbuf->size;
	int need = VARSIZE(tbuf->data) + len;
	if (need < newlen)
		return;
	while (need > newlen)
		newlen *= 2;
	tbuf->data = trealloc(tbuf->data, newlen);
	tbuf->size = newlen;
}

static inline char *get_endp(TBuf *tbuf)
{
	char *p = VARDATA(tbuf->data);
	int len = VARSIZE(tbuf->data) - VARHDRSZ;
	return p + len;
}

static inline void inc_used(TBuf *tbuf, int len)
{
	SET_VARSIZE(tbuf->data, VARSIZE(tbuf->data) + len);
}

static void tbuf_init(TBuf *tbuf, int start_size)
{
	if (start_size < VARHDRSZ)
		start_size = VARHDRSZ;
	tbuf->data = talloc(start_size);
	tbuf->size = start_size;
	SET_VARSIZE(tbuf->data, VARHDRSZ);
}

TBuf *tbuf_alloc(int start_size)
{
	TBuf *res;
	res = talloc(sizeof(TBuf));
	tbuf_init(res, start_size);
	return res;
}

void tbuf_free(TBuf *tbuf)
{
	if (tbuf->data)
		tfree(tbuf->data);
	tfree(tbuf);
}

int tbuf_get_size(TBuf *tbuf)
{
	return VARSIZE(tbuf->data) - VARHDRSZ;
}

void tbuf_reset(TBuf *tbuf)
{
	SET_VARSIZE(tbuf->data, VARHDRSZ);
}

const text *tbuf_look_text(TBuf *tbuf)
{
	return tbuf->data;
}

const char *tbuf_look_cstring(TBuf *tbuf)
{
	char *p;
	request_avail(tbuf, 1);
	p = get_endp(tbuf);
	*p = 0;
	return VARDATA(tbuf->data);
}

void tbuf_append_cstring(TBuf *tbuf, const char *str)
{
	int len = strlen(str);
	request_avail(tbuf, len);
	memcpy(get_endp(tbuf), str, len);
	inc_used(tbuf, len);
}

void tbuf_append_text(TBuf *tbuf, const text *str)
{
	int len = VARSIZE(str) - VARHDRSZ;
	request_avail(tbuf, len);
	memcpy(get_endp(tbuf), VARDATA(str), len);
	inc_used(tbuf, len);
}

void tbuf_append_char(TBuf *tbuf, char chr)
{
	char *p;
	request_avail(tbuf, 1);
	p = get_endp(tbuf);
	*p = chr;
	inc_used(tbuf, 1);
}

text *tbuf_steal_text(TBuf *tbuf)
{
	text *data = tbuf->data;
	tbuf->data = NULL;
	return data;
}

static const char b64tbl[] =
"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
static int b64encode(char *dst, const uint8 *src, int srclen)
{
	char       *p = dst;
	const uint8 *s = src, *end = src + srclen;
	int         pos = 2;
	uint32      buf = 0;

	while (s < end) {
		buf |= (unsigned char) *s << (pos << 3);
		pos--;
		s++;
		/* write it out */
		if (pos < 0) {
			*p++ = b64tbl[ (buf >> 18) & 0x3f ];
			*p++ = b64tbl[ (buf >> 12) & 0x3f ];
			*p++ = b64tbl[ (buf >> 6) & 0x3f ];
			*p++ = b64tbl[ buf & 0x3f ];
			pos = 2;
			buf = 0;
		}
	}
	if (pos != 2) {
		*p++ = b64tbl[ (buf >> 18) & 0x3f ];
		*p++ = b64tbl[ (buf >> 12) & 0x3f ];
		*p++ = (pos == 0) ? b64tbl[ (buf >> 6) & 0x3f ] : '=';
		*p++ = '=';
	}
	return p - dst;
}

static const char hextbl[] = "0123456789abcdef";
static int urlencode(char *dst, const uint8 *src, int srclen)
{
	const uint8 *end = src + srclen;
	char *p = dst;
	while (src < end) {
		if (*src == '=')
			*p++ = '+';
		else if ((*src >= '0' && *src <= '9')
			|| (*src >= 'A' && *src <= 'Z')
			|| (*src >= 'a' && *src <= 'z'))
			*p++ = *src;
		else {
			*p++ = '%';
			*p++ = hextbl[*src >> 4];
			*p++ = hextbl[*src & 15];
		}
	}
	return p - dst;
}

static int quote_literal(char *dst, const uint8 *src, int srclen)
{
	const uint8 *cp1;
	char	   *cp2;
	int			wl;
	bool		is_ext = false;

	cp1 = src;
	cp2 = dst;

	*cp2++ = '\'';
	while (srclen > 0)
	{
		if ((wl = pg_mblen((const char *)cp1)) != 1)
		{
			if (wl > srclen)
				wl = srclen;
			srclen -= wl;

			while (wl-- > 0)
				*cp2++ = *cp1++;
			continue;
		}

		if (*cp1 == '\'') {
			*cp2++ = '\'';
		} else if (*cp1 == '\\') {
			if (!is_ext) {
				memmove(dst + 1, dst, cp2 - dst);
				cp2++;
				is_ext = true;
				*dst = 'E';
			}
			*cp2++ = '\\';
		}
		*cp2++ = *cp1++;
		srclen--;
	}

	*cp2++ = '\'';

	return cp2 - dst;
}

/* check if ident is keyword that needs quoting */
static bool is_keyword(const char *ident)
{
	const ScanKeyword *kw;

	/* do the lookup */
#if PG_VERSION_NUM >= 80500
	kw = ScanKeywordLookup(ident, ScanKeywords, NumScanKeywords);
#else
	kw = ScanKeywordLookup(ident);
#endif

	/* unreserved? */
#if PG_VERSION_NUM >= 80300
	if (kw && kw->category == UNRESERVED_KEYWORD)
		return false;
#endif

	/* found anything? */
	return kw != NULL;
}

/*
 * slon_quote_identifier                     - Quote an identifier only if needed
 *
 * When quotes are needed, we palloc the required space; slightly
 * space-wasteful but well worth it for notational simplicity.
 *
 * Version: pgsql/src/backend/utils/adt/ruleutils.c,v 1.188 2005/01/13 17:19:10
 */
static int
quote_ident(char *dst, const uint8 *src, int srclen)
{
        /*
         * Can avoid quoting if ident starts with a lowercase letter or
         * underscore and contains only lowercase letters, digits, and
         * underscores, *and* is not any SQL keyword.  Otherwise, supply
         * quotes.
         */
        int                     nquotes = 0;
        bool            safe;
        const char *ptr;
        char       *optr;
		char ident[NAMEDATALEN + 1];

		/* expect idents be not bigger than NAMEDATALEN */
		if (srclen > NAMEDATALEN)
			srclen = NAMEDATALEN;
		memcpy(ident, src, srclen);
		ident[srclen] = 0;

        /*
         * would like to use <ctype.h> macros here, but they might yield
         * unwanted locale-specific results...
         */
        safe = ((ident[0] >= 'a' && ident[0] <= 'z') || ident[0] == '_');

        for (ptr = ident; *ptr; ptr++)
        {
                char            ch = *ptr;

                if ((ch >= 'a' && ch <= 'z') ||
                        (ch >= '0' && ch <= '9') ||
                        (ch == '_'))
                        continue; /* okay */

                safe = false;
                if (ch == '"')
                        nquotes++;
        }

        if (safe) {
            if (is_keyword(ident))
                safe = false;
        }

		optr = dst;
		if (!safe)
				*optr++ = '"';

        for (ptr = ident; *ptr; ptr++)
        {
                char            ch = *ptr;

                if (ch == '"')
                        *optr++ = '"';
                *optr++ = ch;
        }
		if (!safe)
				*optr++ = '"';

        return optr - dst;
}


void tbuf_encode_cstring(TBuf *tbuf,
		const char *str,
		const char *encoding)
{
	if (str == NULL)
		elog(ERROR, "tbuf_encode_cstring: NULL");
	tbuf_encode_data(tbuf, (const uint8 *)str, strlen(str), encoding);
}

void tbuf_encode_data(TBuf *tbuf,
		const uint8 *data, int len,
		const char *encoding)
{
	int dlen = 0;
	char *dst;
	if (strcmp(encoding, "url") == 0) {
		request_avail(tbuf, len*3);
		dst = get_endp(tbuf);
		dlen = urlencode(dst, data, len);
	} else if (strcmp(encoding, "base64") == 0) {
		request_avail(tbuf, (len + 2) * 4 / 3);
		dst = get_endp(tbuf);
		dlen = b64encode(dst, data, len);
	} else if (strcmp(encoding, "quote_literal") == 0) {
		request_avail(tbuf, len * 2 + 2);
		dst = get_endp(tbuf);
		dlen = quote_literal(dst, data, len);
	} else if (strcmp(encoding, "quote_ident") == 0) {
		request_avail(tbuf, len * 2 + 2);
		dst = get_endp(tbuf);
		dlen = quote_ident(dst, data, len);
	} else
		elog(ERROR, "bad encoding");
	inc_used(tbuf, dlen);
}

