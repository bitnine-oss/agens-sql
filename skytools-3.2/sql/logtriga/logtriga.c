/* ----------------------------------------------------------------------
 * logtriga.c
 *
 *	  Generic trigger for logging table changes.
 *	  Based on Slony-I log trigger.
 *	  Does not depend on event storage.
 *
 *	Copyright (c) 2003-2006, PostgreSQL Global Development Group
 *	Author: Jan Wieck, Afilias USA INC.
 *
 * Generalized by Marko Kreen.
 * ----------------------------------------------------------------------
 */

#include "postgres.h"

#include "executor/spi.h"
#include "commands/trigger.h"
#include "catalog/pg_operator.h"
#include "catalog/pg_type.h"
#include "utils/typcache.h"
#include "utils/rel.h"

#include "textbuf.h"

PG_FUNCTION_INFO_V1(logtriga);
Datum logtriga(PG_FUNCTION_ARGS);

#ifdef PG_MODULE_MAGIC
PG_MODULE_MAGIC;
#endif

/*
 * There may be several plans to be cached.
 *
 * FIXME: plans are kept in singe-linked list
 * so not very fast access.  Probably they should be
 * handled more intelligently.
 */
typedef struct PlanCache PlanCache;

struct PlanCache {
	PlanCache *next;
	char *query;
	void *plan;
};

/*
 * Cache result allocations.
 */
typedef struct ArgCache
{
	TBuf	   *op_type;
	TBuf       *op_data;
}	ArgCache;


static PlanCache *plan_cache = NULL;
static ArgCache *arg_cache = NULL;

/*
 * Cache helpers
 */

static void *get_plan(const char *query)
{
	PlanCache *c;
	void *plan;
	Oid   plan_types[2];

	for (c = plan_cache; c; c = c->next)
		if (strcmp(query, c->query) == 0)
			return c->plan;

	/*
	 * Plan not cached, prepare new plan then.
	 */
	plan_types[0] = TEXTOID;
	plan_types[1] = TEXTOID;
	plan = SPI_saveplan(SPI_prepare(query, 2, plan_types));
	if (plan == NULL)
		elog(ERROR, "logtriga: SPI_prepare() failed");
	
	/* create cache object */
	c = malloc(sizeof(*c));
	if (!c)
		elog(ERROR, "logtriga: no memory for plan cache");

	c->plan = plan;
	c->query = strdup(query);

	/* insert at start */
	c->next = plan_cache;
	plan_cache = c;

	return plan;
}

static ArgCache *
get_arg_cache(void)
{
	if (arg_cache == NULL) {
		ArgCache *a = malloc(sizeof(*a));
		if (!a)
			elog(ERROR, "logtriga: no memory");
		memset(a, 0, sizeof(*a));
		a->op_type = tbuf_alloc(8);
		a->op_data = tbuf_alloc(8192);
		arg_cache = a;
	}
	return arg_cache;
}

static void
append_key_eq(TBuf *tbuf, const char *col_ident, const char *col_value)
{
	if (col_value == NULL)
		elog(ERROR, "logtriga: Unexpected NULL key value");

	tbuf_encode_cstring(tbuf, col_ident, "quote_ident");
	tbuf_append_char(tbuf, '=');
	tbuf_encode_cstring(tbuf, col_value, "quote_literal");
}

static void
append_normal_eq(TBuf *tbuf, const char *col_ident, const char *col_value)
{
	tbuf_encode_cstring(tbuf, col_ident, "quote_ident");
	tbuf_append_char(tbuf, '=');
	if (col_value != NULL)
		tbuf_encode_cstring(tbuf, col_value, "quote_literal");
	else
		tbuf_append_cstring(tbuf, "NULL");
}

static void process_insert(ArgCache *cs, TriggerData *tg, char *attkind)
{
	HeapTuple	new_row = tg->tg_trigtuple;
	TupleDesc	tupdesc = tg->tg_relation->rd_att;
	int			i;
	int			need_comma = false;
	int			attkind_idx;

	/*
	 * INSERT
	 *
	 * op_type = 'I' op_data = ("non-NULL-col" [, ...]) values ('value' [,
	 * ...])
	 */
	tbuf_append_cstring(cs->op_type, "I");

	/*
	 * Specify all the columns
	 */
	tbuf_append_char(cs->op_data, '(');
	attkind_idx = -1;
	for (i = 0; i < tg->tg_relation->rd_att->natts; i++)
	{
		char *col_ident;

		/* Skip dropped columns */
		if (tupdesc->attrs[i]->attisdropped)
			continue;

		/* Check if allowed by colstring */
		attkind_idx++;
		if (attkind[attkind_idx] == '\0')
			break;
		if (attkind[attkind_idx] == 'i')
			continue;

		if (need_comma)
			tbuf_append_char(cs->op_data, ',');
		else
			need_comma = true;

		/* quote column name */
		col_ident = SPI_fname(tupdesc, i + 1);
		tbuf_encode_cstring(cs->op_data, col_ident, "quote_ident");
	}

	/*
	 * Append the string ") values ("
	 */
	tbuf_append_cstring(cs->op_data, ") values (");

	/*
	 * Append the values
	 */
	need_comma = false;
	attkind_idx = -1;
	for (i = 0; i < tg->tg_relation->rd_att->natts; i++)
	{
		char *col_value;

		/* Skip dropped columns */
		if (tupdesc->attrs[i]->attisdropped)
			continue;

		/* Check if allowed by colstring */
		attkind_idx++;
		if (attkind[attkind_idx] == '\0')
			break;
		if (attkind[attkind_idx] == 'i')
			continue;

		if (need_comma)
			tbuf_append_char(cs->op_data, ',');
		else
			need_comma = true;

		/* quote column value */
		col_value = SPI_getvalue(new_row, tupdesc, i + 1);
		if (col_value == NULL)
			tbuf_append_cstring(cs->op_data, "null");
		else
			tbuf_encode_cstring(cs->op_data, col_value, "quote_literal");
	}

	/*
	 * Terminate and done
	 */
	tbuf_append_char(cs->op_data, ')');
}

static int process_update(ArgCache *cs, TriggerData *tg, char *attkind)
{
	HeapTuple	old_row = tg->tg_trigtuple;
	HeapTuple	new_row = tg->tg_newtuple;
	TupleDesc	tupdesc = tg->tg_relation->rd_att;
	Datum		old_value;
	Datum		new_value;
	bool		old_isnull;
	bool		new_isnull;

	char	   *col_ident;
	char	   *col_value;
	int			i;
	int			need_comma = false;
	int			need_and = false;
	int			attkind_idx;
	int			ignore_count = 0;

	/*
	 * UPDATE
	 *
	 * op_type = 'U' op_data = "col_ident"='value' [, ...] where "pk_ident" =
	 * 'value' [ and ...]
	 */
	tbuf_append_cstring(cs->op_type, "U");

	attkind_idx = -1;
	for (i = 0; i < tg->tg_relation->rd_att->natts; i++)
	{
		/*
		 * Ignore dropped columns
		 */
		if (tupdesc->attrs[i]->attisdropped)
			continue;

		attkind_idx++;
		if (attkind[attkind_idx] == '\0')
			break;

		old_value = SPI_getbinval(old_row, tupdesc, i + 1, &old_isnull);
		new_value = SPI_getbinval(new_row, tupdesc, i + 1, &new_isnull);

		/*
		 * If old and new value are NULL, the column is unchanged
		 */
		if (old_isnull && new_isnull)
			continue;

		/*
		 * If both are NOT NULL, we need to compare the values and skip
		 * setting the column if equal
		 */
		if (!old_isnull && !new_isnull)
		{
			Oid			opr_oid;
			FmgrInfo   *opr_finfo_p;

			/*
			 * Lookup the equal operators function call info using the
			 * typecache if available
			 */
			TypeCacheEntry *type_cache;

			type_cache = lookup_type_cache(SPI_gettypeid(tupdesc, i + 1),
							  TYPECACHE_EQ_OPR | TYPECACHE_EQ_OPR_FINFO);
			opr_oid = type_cache->eq_opr;
			if (opr_oid == ARRAY_EQ_OP)
				opr_oid = InvalidOid;
			else
				opr_finfo_p = &(type_cache->eq_opr_finfo);

			/*
			 * If we have an equal operator, use that to do binary
			 * comparision. Else get the string representation of both
			 * attributes and do string comparision.
			 */
			if (OidIsValid(opr_oid))
			{
				if (DatumGetBool(FunctionCall2(opr_finfo_p,
											   old_value, new_value)))
					continue;
			}
			else
			{
				char	   *old_strval = SPI_getvalue(old_row, tupdesc, i + 1);
				char	   *new_strval = SPI_getvalue(new_row, tupdesc, i + 1);

				if (strcmp(old_strval, new_strval) == 0)
					continue;
			}
		}

		if (attkind[attkind_idx] == 'i')
		{
			/* this change should be ignored */
			ignore_count++;
			continue;
		}

		if (need_comma)
			tbuf_append_char(cs->op_data, ',');
		else
			need_comma = true;

		col_ident = SPI_fname(tupdesc, i + 1);
		col_value = SPI_getvalue(new_row, tupdesc, i + 1);

		append_normal_eq(cs->op_data, col_ident, col_value);
	}

	/*
	 * It can happen that the only UPDATE an application does is to set a
	 * column to the same value again. In that case, we'd end up here with
	 * no columns in the SET clause yet. We add the first key column here
	 * with it's old value to simulate the same for the replication
	 * engine.
	 */
	if (!need_comma)
	{
		/* there was change in ignored columns, skip whole event */
		if (ignore_count > 0)
			return 0;

		for (i = 0, attkind_idx = -1; i < tg->tg_relation->rd_att->natts; i++)
		{
			if (tupdesc->attrs[i]->attisdropped)
				continue;

			attkind_idx++;
			if (attkind[attkind_idx] == 'k')
				break;
		}
		col_ident = SPI_fname(tupdesc, i + 1);
		col_value = SPI_getvalue(old_row, tupdesc, i + 1);

		append_key_eq(cs->op_data, col_ident, col_value);
	}

	tbuf_append_cstring(cs->op_data, " where ");

	for (i = 0, attkind_idx = -1; i < tg->tg_relation->rd_att->natts; i++)
	{
		/*
		 * Ignore dropped columns
		 */
		if (tupdesc->attrs[i]->attisdropped)
			continue;

		attkind_idx++;
		if (attkind[attkind_idx] == '\0')
			break;
		if (attkind[attkind_idx] != 'k')
			continue;
		col_ident = SPI_fname(tupdesc, i + 1);
		col_value = SPI_getvalue(old_row, tupdesc, i + 1);

		if (need_and)
			tbuf_append_cstring(cs->op_data, " and ");
		else
			need_and = true;

		append_key_eq(cs->op_data, col_ident, col_value);
	}
	return 1;
}

static void process_delete(ArgCache *cs, TriggerData *tg, char *attkind)
{
	HeapTuple	old_row = tg->tg_trigtuple;
	TupleDesc	tupdesc = tg->tg_relation->rd_att;
	char	   *col_ident;
	char	   *col_value;
	int			i;
	int			need_and = false;
	int			attkind_idx;

	/*
	 * DELETE
	 *
	 * op_type = 'D' op_data = "pk_ident"='value' [and ...]
	 */
	tbuf_append_cstring(cs->op_type, "D");

	for (i = 0, attkind_idx = -1; i < tg->tg_relation->rd_att->natts; i++)
	{
		if (tupdesc->attrs[i]->attisdropped)
			continue;

		attkind_idx++;
		if (attkind[attkind_idx] == '\0')
			break;
		if (attkind[attkind_idx] != 'k')
			continue;
		col_ident = SPI_fname(tupdesc, i + 1);
		col_value = SPI_getvalue(old_row, tupdesc, i + 1);

		if (need_and)
			tbuf_append_cstring(cs->op_data, " and ");
		else
			need_and = true;

		append_key_eq(cs->op_data, col_ident, col_value);
	}
}

Datum logtriga(PG_FUNCTION_ARGS)
{
	TriggerData *tg;
	Datum		argv[2];
	int			rc;
	ArgCache	*cs;
	TupleDesc	tupdesc;
	int			i;
	int			attcnt;
	char		*attkind;
	char		*kpos;
	char		*query;
	int			need_event = 1;

	/*
	 * Get the trigger call context
	 */
	if (!CALLED_AS_TRIGGER(fcinfo))
		elog(ERROR, "logtriga not called as trigger");
	tg = (TriggerData *) (fcinfo->context);
	tupdesc = tg->tg_relation->rd_att;

	/*
	 * Check all logTrigger() calling conventions
	 */
	if (!TRIGGER_FIRED_AFTER(tg->tg_event))
		elog(ERROR, "logtriga must be fired AFTER");
	if (!TRIGGER_FIRED_FOR_ROW(tg->tg_event))
		elog(ERROR, "logtriga must be fired FOR EACH ROW");
	if (tg->tg_trigger->tgnargs != 2)
		elog(ERROR, "logtriga must be defined with 2 args");

	/*
	 * Connect to the SPI manager
	 */
	if ((rc = SPI_connect()) < 0)
		elog(ERROR, "logtriga: SPI_connect() failed");

	cs = get_arg_cache();

	tbuf_reset(cs->op_type);
	tbuf_reset(cs->op_data);

	/*
	 * Get all the trigger arguments
	 */
	attkind = tg->tg_trigger->tgargs[0];
	query = tg->tg_trigger->tgargs[1];

	/*
	 * Count number of active columns
	 */
	for (i = 0, attcnt = 0; i < tg->tg_relation->rd_att->natts; i++)
	{
		if (tupdesc->attrs[i]->attisdropped)
			continue;
		attcnt++;
	}

	/*
	 * Make sure all 'k' columns exist and there is at least one of them.
	 */
	kpos = strrchr(attkind, 'k');
	if (kpos == NULL)
		elog(ERROR, "logtriga: need at least one key column");
	if (kpos - attkind >= attcnt)
		elog(ERROR, "logtriga: key column does not exist");

	/*
	 * Determine cmdtype and op_data depending on the command type
	 */
	if (TRIGGER_FIRED_BY_INSERT(tg->tg_event))
		process_insert(cs, tg, attkind);
	else if (TRIGGER_FIRED_BY_UPDATE(tg->tg_event))
		need_event = process_update(cs, tg, attkind);
	else if (TRIGGER_FIRED_BY_DELETE(tg->tg_event))
		process_delete(cs, tg, attkind);
	else
		elog(ERROR, "logtriga fired for unhandled event");

	/*
	 * Construct the parameter array and insert the log row.
	 */
	if (need_event)
	{
		argv[0] = PointerGetDatum(tbuf_look_text(cs->op_type));
		argv[1] = PointerGetDatum(tbuf_look_text(cs->op_data));
		SPI_execp(get_plan(query), argv, NULL, 0);
	}
	SPI_finish();
	return PointerGetDatum(NULL);
}

