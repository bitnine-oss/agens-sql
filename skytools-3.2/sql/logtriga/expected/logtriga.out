-- init
\set ECHO none
create table rtest (
	id integer primary key,
	dat text
);
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "rtest_pkey" for table "rtest"
create table clog (
  id serial,
  op text,
  data text
);
NOTICE:  CREATE TABLE will create implicit sequence "clog_id_seq" for serial column "clog.id"
create trigger rtest_triga after insert or update or delete on rtest
for each row execute procedure logtriga('kv',
'insert into clog (op, data) values ($1, $2)');
-- simple test
insert into rtest values (1, 'value1');
update rtest set dat = 'value2';
delete from rtest;
select * from clog; delete from clog;
 id | op |              data              
----+----+--------------------------------
  1 | I  | (id,dat) values ('1','value1')
  2 | U  | dat='value2' where id='1'
  3 | D  | id='1'
(3 rows)

-- test new fields
alter table rtest add column dat2 text;
insert into rtest values (1, 'value1');
update rtest set dat = 'value2';
delete from rtest;
select * from clog; delete from clog;
 id | op |              data              
----+----+--------------------------------
  4 | I  | (id,dat) values ('1','value1')
  5 | U  | dat='value2' where id='1'
  6 | D  | id='1'
(3 rows)

-- test field rename
alter table rtest alter column dat type integer using 0;
insert into rtest values (1, '666', 'newdat');
update rtest set dat = 5;
delete from rtest;
select * from clog; delete from clog;
 id | op |            data             
----+----+-----------------------------
  7 | I  | (id,dat) values ('1','666')
  8 | U  | dat='5' where id='1'
  9 | D  | id='1'
(3 rows)

-- test field ignore
drop trigger rtest_triga on rtest;
create trigger rtest_triga after insert or update or delete on rtest
for each row execute procedure logtriga('kiv',
'insert into clog (op, data) values ($1, $2)');
insert into rtest values (1, '666', 'newdat');
update rtest set dat = 5, dat2 = 'newdat2';
update rtest set dat = 6;
delete from rtest;
select * from clog; delete from clog;
 id | op |              data               
----+----+---------------------------------
 10 | I  | (id,dat2) values ('1','newdat')
 11 | U  | dat2='newdat2' where id='1'
 12 | D  | id='1'
(3 rows)

-- test wrong key
drop trigger rtest_triga on rtest;
create trigger rtest_triga after insert or update or delete on rtest
for each row execute procedure logtriga('vik',
'insert into clog (op, data) values ($1, $2)');
insert into rtest values (1, 0, 'non-null');
insert into rtest values (2, 0, NULL);
update rtest set dat2 = 'non-null2' where id=1;
update rtest set dat2 = NULL where id=1;
update rtest set dat2 = 'new-nonnull' where id=2;
ERROR:  logtriga: Unexpected NULL key value
delete from rtest where id=1;
ERROR:  logtriga: Unexpected NULL key value
delete from rtest where id=2;
ERROR:  logtriga: Unexpected NULL key value
select * from clog; delete from clog;
 id | op |                  data                  
----+----+----------------------------------------
 13 | I  | (id,dat2) values ('1','non-null')
 14 | I  | (id,dat2) values ('2',null)
 15 | U  | dat2='non-null2' where dat2='non-null'
 16 | U  | dat2=NULL where dat2='non-null2'
(4 rows)

